README.md
# Alien Invasion Game - Defend Earth from the Extraterrestrial Onslaught 🚀🛸

![Gameplay Screenshot](screenshot.png)

Welcome to the Alien Invasion Game! This is a Pygame-based project where players defend Earth from an alien invasion. The game features exciting space battles, power-ups, and challenging levels.

## Table of Contents

- [Features](#features)
- [Installation](#installation)
- [Usage](#usage)
- [Controls](#controls)
- [License](#license)
- [Contact](#contact)

## Features

- Multiple levels of increasing difficulty.
- Various types of alien enemies with unique behaviors.
- Player ship with customizable attributes and power-ups.
- Smooth and responsive controls for an engaging gameplay experience.
- Visuals and sound effects to enhance immersion.
- High score tracking to challenge friends and beat your personal best.

## Installation

1. Ensure you have Python [installed](https://www.python.org/downloads/) on your system.
2. Clone this repository: `git clone https://github.com/your-username/alien-invasion-game.git`
3. Navigate to the project directory: `cd alien-invasion-game`
4. Install required dependencies: `pip install pygame`

## Usage

1. Run the game: `python main.py`
2. Use arrow keys to move the player's ship and the "X" or "C" (replenish health) keys to shoot.
3. Defeat waves of alien enemies to progress through levels and earn points.
4. Challenge yourself to achieve the highest score and save Earth from the invasion!

## Controls

- Use the **left arrow** key to move the ship left.
- Use the **right arrow** key to move the ship right.
- Use the **up arrow** key to move the ship up.
- - Use the **up arrow** key to move the ship down.
- Use the **X** or **C** to shoot rockets and destroy alien enemies.

## License

This project is licensed under the [MIT License](LICENSE).

## Contact

Feel free to reach out to me if you have any questions, suggestions, or just want to chat about the game! You can contact me at:

- Email: f.sghiri01@gmail.com


CHANGES.txt
fixed bullet bug:

each t time, alien bullets shoots a new bullet
super bullets currently have no specefic function, you can use just them to restore

updated json manipulation





main.py
import pygame, sys, os, random, json, time, getpass
from Modules import *


# tells the game story
beginning_story()
name = input('ENTER YOUR NAME: ').lower().strip()

#inisialization of pygame: sounds and everything
pygame.init()

#Settings/Stats attributes
AI = Settings() # stands for alien invasion
STATS = Stats(name)


i, data = check_database(STATS)

# screen
SCREEN = AI.screen()
# bg music
AI.bg_sound('Images_and_sounds/space.wav')
# logo
AI.set_logo()
# ships and icons
SHIP = FighterShip(SCREEN, 400, 540, 'Images_and_sounds/sprites/space_ship')
ICON_1 = Ship(SCREEN, 150, 150, 'Images_and_sounds/sprites/icon_1')
ICON_2 = Ship(SCREEN, 400, 300, 'Images_and_sounds/sprites/icon_2')

# buttons
START_BUTTON = Button(SCREEN, 'Images_and_sounds/play_button.png', (200, 90), 400, 500)

#____sprite group____
ICONS = pygame.sprite.Group()
ICONS.add(ICON_1, ICON_2)
# alien ships
ships = pygame.sprite.Group()
#fighter ships
main_ship = pygame.sprite.Group()
main_ship.add(SHIP)

alien_bullets = pygame.sprite.Group()
fighter_bullets = pygame.sprite.Group()

# les compteurs
magazine = 3 # ___limit use of super bullets___
t = 0 # ___manual timer for launching asteroids while in home screen___
a = 0 # ___manual timer for spawning aliens when game begins___

def launch_asteroid():
    '''launches asteroids each t time'''
    global t, ICONS
    t += 1
    if t == 200:
        y = random.randint(50, 550)
        asteroid = Asteroids(SCREEN, 700, y, 'Images_and_sounds/sprites/asteroid')
        ICONS.add(asteroid)
        t = 0

def spawn_ships():
    """creates and spawns alien ships and spawns fighter ship"""
    global a, ships, alien_bullets, SHIP
    a += 1
    if a == 5:
        x = 10
        rand = random.randint(20,750) #random x
        alien_ship = AlienShip(SCREEN, x+rand, 25, 'Images_and_sounds/sprites/aliens', alien_bullets)
        #add to sprites
        ships.add(alien_ship)
        a = -100 # spawn speed
    
    # alien ships
    ships.draw(SCREEN)
    ships.update()

    # alien bullets
    alien_bullets.draw(SCREEN)
    alien_bullets.update()
    
    #__main ship__
    main_ship.draw(SCREEN)
    main_ship.update()

def check_events(button=START_BUTTON):
    '''checks for events when needed: starting quitting moving shooting'''
    global magazine, run, bullets, ships
    # if fighter ship loses 10 health: GAME OVER
    if STATS.health == 0:
        run = True

    #assuring the mouvement of the ship
    pos = pygame.mouse.get_pos()
    for event in pygame.event.get():
        if button.img_rect.collidepoint(pos):  # clicks
            if pygame.mouse.get_pressed()[0] == 1:  # 0 sig right click, 1 sig state: True
                run = False
        if event.type == pygame.QUIT:  # quits
            sys.exit()

        # keydown
        if event.type == pygame.KEYDOWN:
            # quits
            if event.key == pygame.K_ESCAPE:
                sys.exit()
            # enters game with space_bar
            if event.key == pygame.K_SPACE:
                run = False
            # shoots normal bullet
            if event.key == pygame.K_x:
                normal_bullet = Bullet('Images_and_sounds/bomb.png', 'Images_and_sounds/nrml_b.wav', (15, 25), SHIP.rect.centerx, SHIP.rect.centery)
                normal_bullet.sound.play()
                fighter_bullets.add(normal_bullet)
            # shoots super bullet
            if event.key == pygame.K_c and magazine > 0:
                STATS.health = 10
                super_bullet = Bullet('Images_and_sounds/nuclear-bomb.png', 'Images_and_sounds/atomic_b.wav', (17, 25), SHIP.rect.centerx, SHIP.rect.centery)
                super_bullet.sound.play()
                fighter_bullets.add(super_bullet)
                magazine -= 1
    fighter_bullets.draw(SCREEN)
    fighter_bullets.update()

def check_collisions():
    global ships, fighter_bullets, alien_bullets, main_ship, SHIP
    #AUTOMATICALLY REMOVES SHIPS WHEN HIT
    #       true and false sig wether that ship/bullet get cleared or not
    #    removes alien ship when hit add 1 point to score
    if pygame.sprite.groupcollide(ships, fighter_bullets, True, True):
        STATS.points += 1
    #    diminish health when fighter ship gets hit
    if pygame.sprite.groupcollide(main_ship, alien_bullets, False, True): # if alien ship hits main ship, it wont shoot again
        STATS.health -= 1

# WONT BE CALLED: ONLY AFTER ASKING: DELETES OUT OF SCREEN SHIPS AND BULLETS
def refrech():
    ''' refrech game by deleting all non useful variables'''
    global ships, alien_bullets
    # removes out range ships
    for ship in ships:
        if ship.rect.centery > 620 or ship.rect.centery < -10:
            ships.remove(ship)
    # removes out range bullets
    for bullet in alien_bullets:
        if bullet.rect.centery > 1000 or bullet.rect.centery < -10:
            alien_bullets.remove(bullet)

# ___main___
run = True
while run:
    #fixed bg
    SCREEN.blit(AI.bg_img, (0, 0))
    #start button
    START_BUTTON.blit_me()
    #icons
    ICONS.draw(SCREEN)
    ICONS.update()
    #asteroids
    launch_asteroid()
    #check for events
    check_events()

    #updates display to newest img
    pygame.display.flip()


while not run:
    #animated screen
    AI.animated_bg(SCREEN)

    #fighter ship and alien ship: mvt and shoot
    spawn_ships()

    # check for events and collisions
    check_events()
    check_collisions()

    #prints health and super bullets left
    health_left = write('health: ' + str(STATS.health))
    magazine_left = write('super bullets left: ' + str(magazine))
    score = write('score: ' + str(STATS.points))
    SCREEN.blit(score, (650, 550))
    SCREEN.blit(magazine_left, (10, 550))
    SCREEN.blit(health_left, (10, 520))

    #updates display to newest img and refrech game
    refrech()
    pygame.display.flip()


game_over = write('Game Over')
score = write('score: ' + str(STATS.points))
congrats_msg = update_database(STATS, data, i)
congrats_text = write(congrats_msg, 12)

while run:
    SCREEN.blit(AI.bg_img, (0, 0))
    #prints game over
    SCREEN.blit(game_over, (320, 270))
    #prints score
    if congrats_msg:
        SCREEN.blit(congrats_text, (250, 400))
    else:
        SCREEN.blit(score, (320, 330))
    #checks...
    check_events()
    pygame.display.update()

"""
Firas Sghiri...!
"""


Modules.py
import pygame, sys, os, random, time, json, getpass
pygame.init() # for the fonts

# classes for basic game functions
class Settings:
    '''sets the screen and its basic parts'''
    def __init__(self):
        self.logo_img = pygame.image.load('Images_and_sounds/logo_ufo.png')
        self.bg_img = pygame.image.load('Images_and_sounds/bg.png')
        self.bg_img = pygame.transform.scale(self.bg_img, (800, 600))

        #y cordinate counter for bg
        self.y = 0

    def screen(self):
        """sets the screen settings"""
        screen = pygame.display.set_mode((800, 600))
        pygame.display.set_caption('Alien Invasion')
        return screen

    def set_logo(self):
        """sets the logo"""
        pygame.display.set_icon(self.logo_img)

    def animated_bg(self, screen):
        '''animates the bg img'''
        screen.blit(self.bg_img, (0, self.y))  # first rolling image
        screen.blit(self.bg_img, (0, -600 + self.y))  # second rolling image
        self.y += 0.5
        if self.y == 600:  # loop again
            self.y = 0

    def bg_sound(self, file_path):
        """runs sus bg music"""
        pygame.mixer.init()
        pygame.mixer.music.load(file_path)
        pygame.mixer.music.play(-1)


class Stats:
    '''for stats like ship health, player score, etc...'''
    def __init__(self, name):
        self.name = name
        self.health = 10
        self.points = 0


# ship-type classes
class Ship(pygame.sprite.Sprite):
    """sets all ship kind of objects"""
    def __init__(self, screen, pos_x, pos_y, path):
        """Initialize the ship and set its starting position and mouvements and shoots."""
        super().__init__()
        self.screen = screen

        # sprites
        self.sprites = []
        #   looping through files
        for file in os.listdir(path):
            self.sprites.append(pygame.image.load(path+'/'+file))

        self.current_sprite = 0
        self.image = self.sprites[int(self.current_sprite)]
        self.rect = self.image.get_rect()
        self.rect.centerx, self.rect.centery = pos_x, pos_y

    def update(self):
        # animation
        self.current_sprite += 0.1
        if int(self.current_sprite) == len(self.sprites):
            self.current_sprite = 0
        self.image = self.sprites[int(self.current_sprite)]


class Asteroids(Ship):
    '''ship type = asteroid'''
    def __init__(self, screen, pos_x, pos_y, path):
        super().__init__(screen, pos_x, pos_y, path)

    def update(self):
        # animation
        self.current_sprite += 0.1
        if int(self.current_sprite) == len(self.sprites):
            self.current_sprite = 0
        self.image = self.sprites[int(self.current_sprite)]

        # motion for asteroids
        self.rect.centerx -= 2
        self.rect.centery += 1


class FighterShip(Ship):
    '''ship type = Fighting ship'''
    def __init__(self, screen, pos_x, pos_y, path):
        super().__init__(screen, pos_x, pos_y, path)

    def update(self):
        #animation
        self.current_sprite += 0.1
        if int(self.current_sprite) == len(self.sprites):
            self.current_sprite = 0
        self.image = self.sprites[int(self.current_sprite)]

        #mouvement
        key = pygame.key.get_pressed()
        # setting edges: if it gets out it comes back to starting position
        if 0 < self.rect.centery < 600 and 0 < self.rect.centerx < 800:
            if key[pygame.K_UP]:
                self.rect.centery -= 3
            if key[pygame.K_DOWN]:
                self.rect.centery += 3
            if key[pygame.K_LEFT]:
                self.rect.centerx -= 3
            if key[pygame.K_RIGHT]:
                self.rect.centerx += 3
        else:
            self.rect.centerx, self.rect.centery = 400, 540  # reset x,y cordinates
        
        #shooting
        """in the check events in the main file"""


class AlienShip(Ship):
    '''ship type = alien ship'''
    def __init__(self, screen, pos_x, pos_y, path, bullet_group):
        super().__init__(screen, pos_x, pos_y, path)
        
        self.alien_bullets = bullet_group
        self.counter = 0
        self.bullet_counter = 90

    def update(self):
        #vertival descending
        self.rect.centery += 1
        # surprise teleportation on x axis and shooting
        self.counter += 1
        rand_x = random.randint(-50, 50)
        if self.counter == 200:
            if 20 < (self.rect.centerx + rand_x) < 760:
                self.rect.centerx += rand_x
                self.counter = 0
                
        # shoot bullets
        self.bullet_counter += 1
        if self.bullet_counter == 100:
            self.bullet_counter = 0
            # create bullet
            alien_bullet = AlienBullet('Images_and_sounds/alien_bullet.png', 'Images_and_sounds/alien_b.wav', (5, 15), self.rect.centerx, self.rect.centery)
            # play shooting sound
            alien_bullet.sound.play()
            # add to sprites
            self.alien_bullets.add(alien_bullet)



# bullet-type clsses
class Bullet(pygame.sprite.Sprite):
    """sets all bullet kind of objects"""
    def __init__(self, img_file_path, sound_file_path, size, pos_x, pos_y):
        super().__init__()

        # bullet img, sound and rect
        self.image = pygame.image.load(img_file_path)
        self.image = pygame.transform.scale(self.image, size)

        #ship cords
        self.x = pos_x
        self.y = pos_y

        self.rect = self.image.get_rect()
        self.rect.centerx, self.rect.y = self.x, self.y

        self.sound = pygame.mixer.Sound(sound_file_path)

    def update(self):
        self.rect.centery -= 5 #bullet speed
        if self.rect.centery < -10 : # if bullet passes by y = -10, stop it
            self.rect.centery = -10


class AlienBullet(Bullet):
    '''bullet type = those coming from alien ships'''
    def __init__(self, img_file_path, sound_file_path, size, pos_x, pos_y):
        super().__init__(img_file_path, sound_file_path, size, pos_x, pos_y)

    def update(self):
        self.rect.centery += 5 # bullet speed


# button type classes
class Button:
    '''sets buttons'''
    def __init__(self, screen, img_file, size, x_pos, y_pos):
        self.img = pygame.image.load(img_file)
        self.img = pygame.transform.scale(self.img, size)
        self.img_rect = self.img.get_rect()
        self.img_rect.centerx, self.img_rect.centery = x_pos, y_pos
        self.screen = screen

    def blit_me(self):
        self.screen.blit(self.img, self.img_rect)

#functions
def write(msg, z=30):
    # write message
    font = pygame.font.SysFont('Comic Sans MS', z)
    text = font.render(msg, False, (255, 54, 61))
    return text

def beginning_story():
    time.sleep(1)
    print('\n')
    print('\n')
    print('\n')
    print('\tNOTE: pass is 13X + 23 = 920')
    time.sleep(1)
    print('YOU ARE IN FRONT OF A SHIP DOOR')
    time.sleep(1.5)
    print('\tDOOR REQUIRES PASSWORD...')
    time.sleep(1)
    # door password
    password = getpass.getpass('type door pass:')
    if password != '69':
        print('YOU FELL IN A TRAP!!!')
        time.sleep(1)
        print('FALLING...')
        time.sleep(1)
        print('YOU DIED FROM FALL')
        time.sleep(1)
        sys.exit()
    skip = input('you want to know the story? (y/n) ').lower().strip()
    if skip == 'y':
        # entering the ship
        print('\tyear is 3500...')
        time.sleep(1)
        print('\tWELCOME TO THE SHIP')
        time.sleep(2)
        answer = input('\t...you got a message, answer? (y/n) ').lower().strip()
        if answer == 'y':
            # the call starts
            print('[anonym]: my salutes to the pilot.')
            time.sleep(1)
            print('[anonym]: some weird creatures from outer space tried to take down Earth!!')
            time.sleep(3)
            print('[anonym]: only a couple of hundreds of people survived the attack, thankfully.')
            time.sleep(3)
            print('[anonym]: bu- b-..., wait a sec i hear a noise!')
            time.sleep(1.5)
            print('\t(went to check)')
            time.sleep(5)
            print('[anonym]: OH SH*T, its time for me to go..')
            time.sleep(2)
            print('[anonym]: I am one of the survivers, i crea- (coughs)!')
            time.sleep(2)
            print('[anonym]: i created this ship so that we-')
            time.sleep(2)
            print('\t[anonym]: -coughing harder-')
            time.sleep(3)
            print('[anonym]: so that we can stop this attack before it even happens!')
            time.sleep(3)
            print('[anonym]: this is our last-..')
            time.sleep(1)
            # call ends
            print('\tBATTERY DIED!')
            time.sleep(0.5)
            print('\tCLOSING CALL!')
            time.sleep(0.5)
            print('\tGame Starting...')
            time.sleep(0.5)
        else:
            # quit story if error
            print('\tGame Starting...')
            time.sleep(2)
    else:
        time.sleep(0.5)
        print('\tGame Starting...')
        time.sleep(0.5)

# database related
def search(name, data):
    """search name in database"""
    for i in range(len(data["players"])):
        if name in data["players"][i]:
            return i
    else:
        return 0

def check_database(STATS):
    """json manipulation"""

    with open('Alien_Invasion.json', 'r+') as file:
        data = json.load(file)
    
    i = search(STATS.name, data)
    if i:
        time.sleep(0.5)
        print('\tGood to see you again ' + STATS.name.title() + '!')
        time.sleep(0.5)
    return i, data

def update_database(STATS, data, i):
    if i: 
        if data["players"][i][STATS.name] < STATS.points < data["record"]: # if new score is better than old take it
            data["players"][i][STATS.name] = STATS.points
            
            with open('Alien_Invasion.json', 'r+') as file:
                file.seek(0)
                json.dump(data, file, indent = 4)

            return "u re getting better"
    
    else:
        with open('Alien_Invasion.json', 'r+') as file:
            data["players"].append({STATS.name: STATS.points})
            file.seek(0)
            json.dump(data, file, indent = 4)
    
    if STATS.points > data["record"]: # checks for new record
        data["record"] = STATS.points
        data["holder"] = STATS.name

        with open('Alien_Invasion.json', 'r+') as file:
            file.seek(0)
            json.dump(data, file, indent = 4)
        
        return f'Congrats {STATS.name} you beat the record holder {data["holder"]}: {data["record"]}'
